Source/Dafny/Compiler-js.cs

```diff
@@ -1042,9 +1042,51 @@ protected class ClassWriter : IClassWriter {
     // ----- Statements -------------------------------------------------------------
 
     protected override void EmitPrintStmt(TargetWriter wr, Expression arg) {
-      wr.Write("process.stdout.write(_dafny.toString(");
-      TrExpr(arg, wr, false);
-      wr.WriteLine("));");
+      bool isString = arg.Type.AsCollectionType != null &&
+                      arg.Type.AsCollectionType.AsSeqType != null &&
+                      arg.Type.AsCollectionType.AsSeqType.Arg.IsCharType;
+      bool isStringLiteral = arg is StringLiteralExpr;
+      bool isGeneric = arg.Type.AsCollectionType != null &&
+                       arg.Type.AsCollectionType.AsSeqType != null &&
+                       arg.Type.AsCollectionType.AsSeqType.Arg.IsTypeParameter;
+      if (isString && isStringLiteral) {
+        // process.stdout.write(_dafny.toString(x));
+        wr.Write("process.stdout.write(_dafny.toString(");
+        TrExpr(arg, wr, false);
+        wr.WriteLine("));");
+      } else if (isString && !isStringLiteral) {
+        // try { process.stdout.write(_dafny.toString((x).join(""))); } catch (_error) { process.stdout.write(_dafny.toString(x)); }
```

Comment:
The `try`-`catch` encloses the `write`, `toString`, and `join`, but I think it's intended just for the `join`. I suggest adding a function to the JavaScript runtime that takes `x` as a parameter and returns `x.join("")` if possible and `x` otherwise. Then, the line emitted here would be just

    process.stdout.write(_dafny.toString(NEWFUNCTION(x)))

This also leaves me wondering, if `x` is a string, why would `x.join("")` fail in the first place?

For that matter, why just have a new routine `_dafny.seqAsString` that does the equivalent of `_dafny.toString(x.join(""))`, and possibly without needing to do the `.join("")` in the first place?

Reply:
join() can fail because, in at least the case of the argument being something like "a"+"b",
the Dafny type of the argument is a seq<char>, but the translated JS evaluates to a literal string, instead of to a sequence.  So the try-catch is a catch-all, rather than trying to predict the exact behavior of arbitrary arguments.

Reply:
I suppose one could argue that Seq.Concat in DafnyRuntime.js should not return a+b, but rather a Sequence equivalent, but that would have left me to find all other such situations.

Reply:
Created the suggested helper function